### PR TITLE
Store message clients in the database.

### PR DESCRIPTION
**Pull request by lamara**
Storing client options in the database now -- added an "id" field to clients that can be configured, to ensure uniqueness.

We are storing the client's private key in the DB -- is this ok?

@mattcollier I removed the MessageClient class -- a note said it was deprecated so I just changed us to the configure only option and updated dependencies accordingly -- there aren't any traces of it anymore, that ok?

### CODE REVIEW

**Comment by mattcollier** [comment]
Don't see any issues here.  We will need to find a private key protection strategy at some point.  I assume storing the private key in the database was part of the specification for now.  As for the MessageClient, I guess we just need to update the modules that were using it to do it the new way?

**Comment by lamara** [comment]
The only module using it was truecred-idp -- already updated it, can push that up to one of our dev branches.

**Comment by lamara** [comment]
Going to use the PR to move over to storing our keypairs in brKey, and just storing a public key identifier for the keypair in the client options here, talked with @dlongley about it.

**Comment by lamara** [comment]
truecred integration here: https://github.com/accreditrust/truecred-idp/pull/13

**Comment by dlongley** [comment]
I want to discuss other options vs. storing private key in DB.

**Comment by dlongley** [comment]
I think what we'll do is add a TODO indicating that we want to use `bedrock-ssm-client` for signing in the future and add an issue for that as well. Then merge this so we can use it in POCs.

File: lib/client.js [hunk 1]
```diff
@@ -251,3 +281,113 @@ function pollServer(options, callback) {
...
+  async.auto({
+    checkPermission: function(callback) {
+      brIdentity.checkPermission(
+        actor, PERMISSIONS.MESSAGE_CLIENT_ADMIN, callback);
+    },
+    update: ['checkPermission', function(callback) {
+      // build a database update
+      var update = database.buildUpdate(client, 'client', {
+        include: [
+          'client.id',
```
Comment (dlongley) [inline lib/client.js]: Can you really change the client ID? We usually don't permit this.

File: lib/client.js [hunk 2]
```diff
@@ -108,86 +165,59 @@ api.start = function(endpoint, interval, callback) {
...
+            ));
+          }
+          callback(err, privateKey);
+        });
+    },
+    poll: ['getPrivateKey', function(callback, results) {
+      console.log("Private key");
+      console.log(results.getPrivateKey);
+      console.log("Public key id");
+      console.log(client.publicKeyId);
```
Comment (dlongley) [inline lib/client.js]: Remove `console.log`s.

File: lib/client.js [hunk 3]
```diff
@@ -108,86 +165,59 @@ api.start = function(endpoint, interval, callback) {
...
     }
   };
   async.auto({
-    poll: function(callback) {
+    getPrivateKey: function(callback) {
+      brKey.getPublicKey({id: client.publicKeyId}, null,
+        function(err, publicKey, meta, privateKey) {
+          if(!privateKey) {
+            return callback(new BedrockError(
+            'Contacting server with no private key configured: ' +
```
Comment (dlongley) [inline lib/client.js]: Fix indent.

File: lib/client.js [hunk 4]
```diff
@@ -108,86 +165,59 @@ api.start = function(endpoint, interval, callback) {
...
     type: 'messageClient.failedToRetrieveMessage',
     details: {
-      endpoint: options.endpoint
+      endpoint: client.endpoint
     }
   };
   async.auto({
-    poll: function(callback) {
+    getPrivateKey: function(callback) {
+      brKey.getPublicKey({id: client.publicKeyId}, null,
```
Comment (dlongley) [inline lib/client.js]: Add TODO for using SSM/HSM.

File: lib/client.js [hunk 5]
```diff
@@ -108,86 +165,59 @@ api.start = function(endpoint, interval, callback) {
...
-  });
-};
-
 // exposed for testing
 api._pollServer = function(options, callback) {
   pollServer(options, callback);
 };
 
 function pollServer(options, callback) {
+  console.log("Poll server called with options", options);
```
Comment (dlongley) [inline lib/client.js]: Remove `console.log`.

File: lib/client.js [hunk 6]
```diff
@@ -28,39 +35,82 @@ bedrock.events.on('bedrock.test.configure', function() {
...
+    }],
+    startClients: ['getClients', function(callback, results) {
+      var clients = results.getClients;
+      console.log("Got clients, defining", clients);
+      async.each(clients, function(client, callback) {
+        var jobId = createJobId(client);
+        var endpoint = client.endpoint;
+        var privateKeyPem = client.privateKeyPem;
+        var publicKeyId = client.publicKeyId;
+        var strictSSL = !('strictSSL' in client) ?
```
Comment (dlongley) [inline lib/client.js]: Remove negation and flip last two parts in ternary.

File: lib/client.js [hunk 7]
```diff
@@ -28,39 +35,82 @@ bedrock.events.on('bedrock.test.configure', function() {
...
         });
-      };
+      }, callback);
+    }],
+    getClients: ['insertClients', function(callback, results) {
+      api.getAll(null, callback);
+    }],
+    startClients: ['getClients', function(callback, results) {
+      var clients = results.getClients;
+      console.log("Got clients, defining", clients);
```
Comment (dlongley) [inline lib/client.js]: Remove `console.log`.

**Pull request closed without merging**